### ISSUE TITLE
Explore consistent kubeconfig settings between Tekton and K8s ext

### ISSUE THREAD

**Issue by siamaksade** [redhat-developer/vscode-tekton#312]
The Kubernetes ext allows defining a custom kubeconfig. When user does that, the Tekton ext and Kubernetes ext would display different clusters which creates problems when user uses Kubernetes actions to create a pipeline via YAML and that ends up on the wrong cluster/namespace.

**Comment by evidolob** [comment]
I have done an investigation and found next:
`vscode-kubernetos` extension provide current k8s configuration, but not provide event when it changed. But `vscode-kubernetos` store path to current configuration in vscode preferences, we in tekton extension can read and receive event when it change. So we able to track when user start using another configuration file. 
To finish this issue I have two options:

1. Integrate deeply with  `vscode-kubernetos` extension, use their api to invoke `kubectl`,  in this case we automatically pick up changes, and need to rewrite the way how we use `tkn`, by adding path to current configuration file.

2. Do not integrate `vscode-kubernetos`, just check if it installed and their configuration, and pass that configuration to all invocations of `kubectl` and `tkn` cli.

Basically both of this options requires almost same amount of work, so design should be more strategic/architectural.

@siamaksade @jeffmaury WDYT?

**Comment by jeffmaury** [comment]
Is an extension allowed to read preferences declared by other extensions ? I thought it was not possible

**Comment by evidolob** [comment]
Hm, I try and it work, at least in dev mode, you just need to know configuration keys

**Comment by gorkem** [comment]
Can we have a 3rd alternate where `vscode-kubernetes-tools` set the environment variable like KUBECONFIG=$KUBECONFIG:/path/to/new/kubeconfig? I think all of our tools should be able to pick up this change, (if not they should). Also nowadays it should be able to set env variable to the integrated terminal too as an added bonus.

**Comment by evidolob** [comment]
@gorkem Yes, I think it an option, but now they not do it, they just write path to preferences from user input: in https://github.com/Azure/vscode-kubernetes-tools/blob/e692354159d038d308546a295703984e7be5b1fe/src/extension.ts#L1925 
`setActiveKubeconfig` just write new value in properties. 
I can make a PR to fix that

**Comment by lstocchi** [comment]
@evidolob look at https://github.com/Azure/vscode-kubernetes-tools/blob/e692354159d038d308546a295703984e7be5b1fe/src/components/kubectl/kubeconfig.ts#L39 
It first checks if the property is set, if nothing is found, it checks other things. If you only listens to that property you won't have a valid result. I would vote for making the k8s tools send info to external extensions when the context changes.

**Comment by evidolob** [comment]
@lstocchi I agree with you, I never say that it is the best solution, I want to use that method to get current config, and use 
```typescript
 vscode.workspace.onDidChangeConfiguration(e => {
    if (e.affectsConfiguration('vs-kubernetes',)) {
       //refresh tekton tree there
    }
  });
```
I cannot find better way to get notification that something changed in k8s extension, without k8s extension modification.
And it is definitely better to improve  k8s extension, instead of making a hack in tekton extension

**Comment by siamaksade** [comment]
> Can we have a 3rd alternate where 
Is it relevant to send a PR to k8s ext switch to using `KUBECONFIG` env var? Otherwise I prefer (2)

**Comment by evidolob** [comment]
> Is it relevant to send a PR to k8s ext switch to using `KUBECONFIG` env var? Otherwise I prefer (2)
@siamaksade Yes, to have `KUBECONFIG` env var set up, we need to send PR for k8s exttension

**Comment by gorkem** [comment]
Just that you are aware this is not a problem unique to tekton. `vscode-knative` and `vscode-openshift-tools` also does similar things. The root cause of the problem is `vscode-kubernetes-tools` has a feature that does something unexpected with kubecontext so the solution should start from `vscode-kubernetes-tools` and we adjust the other extension. 
cc @dgolovin @joshuawilson

**Comment by evidolob** [comment]
OK, so what I should do with this? I may(and want) to make PR for `vscode-kubernetes-tools` with: 

- expose to `KUBECONFIG` env variable path to current config file
- set that variable to integrated terminal
- add some sort of notification for  `vscode-kubernetes-tools` API to inform other extensions that path to config file is changed.

And after that returns to this issue for `vscode-tekton`. I ideal case it will require just handling that event and refresh tekton tree.

Of course if no one objects, or maybe have a better plan.

**Comment by gorkem** [comment]
I would start with 
>     * add some sort of notification for  `vscode-kubernetes-tools` API to inform other extensions that path to config file is changed.

The other 2 are nice to haves that does not solve the original problem

**Comment by joshuawilson** [comment]
`vscode-knative` watches the files listed in `KUBECONFIG` ENV and the `.kube/` so that if there is a change it refreshes to pick up the difference. 

But I agree, having `vscode-kubernetes-tools` own the notification out when it changes makes sense.

**Comment by evidolob** [comment]
@gorkem @lstocchi 
I have PR with events: https://github.com/Azure/vscode-kubernetes-tools/pull/766

**Comment by evidolob** [comment]
Azure/vscode-kubernetes-tools#766 is finally merged(thx @lstocchi for assist), we can continue with this after `vscode-kubernetes-tools` next release.
